driver-core/src/main/com/mongodb/event/ConnectionPoolListener.java

```diff
@@ -20,6 +20,7 @@
 
 /**
  * A listener for connection pool-related events.
+ * Methods of this interface must not throw {@link Exception}s.
```

Comment:
> This interface is meant to be implemented by applications.

Indeed, I did not think about it. In this case we cannot make such a specification change because it may render some user implementations incorrect.

This brings me back to the proposal [should use `try-finally`](https://docs.google.com/document/d/1w3qQb6SVbCNwkQqZvqFjg-JOaO38GhKacoKGyfre2kA/edit#heading=h.w32ub1yqz8uh). I started marking some methods with _"Must not throw `Exception`s"_ only to make sense of the existing code and to write new code that more or less follows the existing style, which does not always use `try-finally` to ensure that operations are executed (e.g., that a callback is completed). The current state of the codebase is the reason I did not argue with @rozza when he asked me to remove `try-finally` blocks.

1. [methods marked as not throwing exceptions] The approach of marking methods as not throwing exceptions and then relying on that when omitting `try-finally` is brittle (it may and likely will be broken by future code changes) and very difficult to use in practice, because it requires constantly thinking about which methods may throw an exception and checking their docs.
2. [`try-finally` used always] After thinking even more, I must admit that it seems impossible to write readable code where `try-finally` applied every time we may not complete a callback due to a method throwing an exception. The reason is that we do not allow completing a callback multiple times (cannot find the code right now, but at least some tests supply callbacks that fail on purpose if completed multiple times). This prevents us from following the usual style that easily handles all exceptions:
```java
```

Reply:
I reverted this change.

Reply:
Thanks for reverting.  I'm a bit surprised that we don't do this already, but it would probably be wise to wrap all calls to event handlers in try catch blocks.  This is already done when multiple listeners are configured (see `EventListenerHelper#getConnectionPoolListener` and `ConnectionPoolListenerMulticaster`), but not, it seems, when there is a single listener.